### PR TITLE
ci: use go.mod to determine go version

### PR DESCRIPTION
**Pull request by aeddi**
Related to https://github.com/gnolang/gno/pull/3229#discussion_r1862370753

This PR replaces most fixed versions of Go in the CI workflows with retrieving the version from the relevant go.mod file. 

For workflows that do not have an associated go.mod file, the go.mod file at the root of the repository is used. All `*_template.yml` workflows seem designed to use a fixed version of go and do not allow passing a go.mod file to the `setup-go` action. Achieving this would require a more significant refactor.

<details><summary>Contributors' checklist...</summary>

- [ ] Added new tests, or not needed, or not feasible
- [ ] Provided an example (e.g. screenshot) to aid review or the PR is self-explanatory
- [ ] Updated the official documentation or not needed
- [ ] No breaking changes were made, or a `BREAKING CHANGE: xxx` message was included in the description
- [ ] Added references to related issues and PRs
- [ ] Provided any useful hints for running manual tests
</details>

### CODE REVIEW

**Comment by Gno2D2** [comment]
I'm a bot that assists the Gno Core team in maintaining this repository. My role is to ensure that contributors understand and follow our guidelines, helping to streamline the development process.

The following requirements must be fulfilled before a pull request can be merged.
Some requirement checks are automated and can be verified by the CI, while others need manual verification by a staff member.

These requirements are defined in this [configuration file](https://github.com/gnolang/gno/tree/master/contribs/github-bot/internal/config/config.go).

## Automated Checks

 🟢 Maintainers must be able to edit this pull request
 🟢 The pull request head branch must be up-to-date with its base


## Manual Checks

- [ ] Determine if infra needs to be updated before merging


<details><summary><b>Debug</b></summary><blockquote>
<details><summary><b>Automated Checks</b></summary><blockquote>

<details><summary>Maintainers must be able to edit this pull request</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 On every pull request

```
### Then
```
🟢 Requirement satisfied
└── 🟢 Maintainer can modify this pull request

```
</blockquote></details>

<details><summary>The pull request head branch must be up-to-date with its base</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 On every pull request

```
### Then
```
🟢 Requirement satisfied
└── 🟢 Head branch (aeddi:dev/aeddi/ci-go-version-file) is up to date with base (master): behind by 0 / ahead by 1

```
</blockquote></details>

</blockquote></details>


<details><summary><b>Manual Checks</b></summary><blockquote>

<details><summary>Determine if infra needs to be updated before merging</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 And
    ├── 🟢 The base branch matches this pattern: master
    └── 🟢 Or
        ├── 🔴 A changed file matches this pattern: Dockerfile
        ├── 🔴 A changed file matches this pattern: ^misc/deployments
        ├── 🔴 A changed file matches this pattern: ^misc/docker-
        ├── 🟢 A changed file matches this pattern: ^.github/workflows/releaser.*\.yml$ (filename: .github/workflows/releaser-master.yml)
        └── 🔴 A changed file matches this pattern: ^.github/workflows/portal-loop\.yml$

```
### Can be checked by
 - team devops

</blockquote></details>

</blockquote></details>

</blockquote></details>

**Review by n0izn0iz — APPROVED** [review]
one `go.mod` to rule them all

**Review by thehowl — COMMENTED** [review]
Noice

**Review by thehowl — APPROVED** [review]

**Pull request merged**
